# Analyze NFL Play Type by team

<font color = red > The goal of the analysis will be to analyze the play calling of the Kansas City Chiefs football team. 

1. Download 2018 nfl play calling data from github
2. Review data to identify important data fields for analysis
3. Analyze pass play tendancy based on the down and distance of the KC Chiefs vs all NFL teams
4. Determine which NFL teams have the most pass-heavy play calling  


 </font>

## Step 1: download nfl play calling data from github

In [5]:
import pandas as pd # data manipulation library
import numpy as np # numerical computation library
import datetime as dt 

import matplotlib.pyplot as plt # plotting library
from matplotlib import cm # color maps for plotting
plt.style.use('ggplot') # use the ggplot plotting style


nfl = pd.read_csv('https://raw.githubusercontent.com/ryurko/nflscrapR-data/master/play_by_play_data/regular_season/reg_pbp_2018.csv', low_memory=False,error_bad_lines=False)





## Step 2: Review data to identify important data fields for analysis

In [6]:
bins = [0, 2, 5, 9.99,10, 15, np.inf]
names = ['<2', '2-5', '6-9', '10', '11-14','15+']

nfl['ydstogo_label'] = pd.cut(nfl['ydstogo'], bins=bins, labels=names)

chiefs = nfl.loc[nfl['posteam'] == "KC"]

In [7]:
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)

In [8]:
nfl.head()

play_id     game_id home_team away_team posteam posteam_type defteam  \
0       37  2018090600       PHI       ATL     ATL         away     PHI   
1       52  2018090600       PHI       ATL     ATL         away     PHI   
2       75  2018090600       PHI       ATL     ATL         away     PHI   
3      104  2018090600       PHI       ATL     ATL         away     PHI   
4      125  2018090600       PHI       ATL     ATL         away     PHI   

  side_of_field  yardline_100   game_date  quarter_seconds_remaining  \
0           PHI          35.0  2018-09-06                      900.0   
1           ATL          75.0  2018-09-06                      900.0   
2           ATL          80.0  2018-09-06                      900.0   
3           ATL          70.0  2018-09-06                      862.0   
4           ATL          59.0  2018-09-06                      826.0   

   half_seconds_remaining  game_seconds_remaining game_half  quarter_end  \
0                  1800.0                  3600.0     Half1            0   
1                  1800.0                  3600.0     Half1            0   
2                  1800.0                  3600.0     Half1            0   
3                  1762.0                  3562.0     Half1            0   
4                  1726.0                  3526.0     Half1            0   

   drive  sp  qtr  down  goal_to_go      time   yrdln  ydstogo  ydsnet  \
0      1   0    1   NaN         0.0  15:00:00  PHI 35        0      73   
1      1   0    1   1.0         0.0  15:00:00  ATL 25       10      73   
2      1   0    1   1.0         0.0  15:00:00  ATL 20       15      73   
3      1   0    1   2.0         0.0  14:22:00  ATL 30        5      73   
4      1   0    1   1.0         0.0  13:46:00  ATL 41       10      73   

                                                desc play_type  yards_gained  \
0  J.Elliott kicks 65 yards from PHI 35 to end zo...   kickoff             0   
1  (15:00) PENALTY on ATL-L.Paulsen, False Start,...   no_play             0   
2  (15:00) M.Ryan pass short right to J.Jones pus...      pass            10   
3  (14:22) J.Jones left end pushed ob at ATL 41 f...       run            11   
4  (13:46) D.Freeman right end to PHI 39 for 20 y...       run            20   

   shotgun  no_huddle  qb_dropback  qb_kneel  qb_spike  qb_scramble  \
0        0          0          0.0         0         0            0   
1        0          0          0.0         0         0            0   
2        0          0          1.0         0         0            0   
3        0          0          0.0         0         0            0   
4        0          0          0.0         0         0            0   

  pass_length pass_location  air_yards  yards_after_catch run_location  \
0         NaN           NaN        NaN                NaN          NaN   
1         NaN           NaN        NaN                NaN          NaN   
2       short         right        8.0                2.0          NaN   
3         NaN           NaN        NaN                NaN         left   
4         NaN           NaN        NaN                NaN        right   

  run_gap field_goal_result  kick_distance extra_point_result  \
0     NaN               NaN            NaN                NaN   
1     NaN               NaN            NaN                NaN   
2     NaN               NaN            NaN                NaN   
3     end               NaN            NaN                NaN   
4     end               NaN            NaN                NaN   

  two_point_conv_result  home_timeouts_remaining  away_timeouts_remaining  \
0                   NaN                        3                        3   
1                   NaN                        3                        3   
2                   NaN                        3                        3   
3                   NaN                        3                        3   
4                   NaN                        3                        3   

   timeout

useful columns appear to be as follows: 

| Column | Description |
| --------------- | ----------- |
| ydstogo_label | e.g. <2, 2-5, 5-8, 8-9, 10 10-15, 15+|
| play_type	 | no_play, pass, kickoff, run|
| down | NaN, 0, 1, 2, 3,4|
| yardline_100	 | 1-99 with numbers > 50 on home side of field |
|posteam | team with possession|

## Step 3: Analyze pass play tendancy of the KC Chiefs vs all NFL teams

In [9]:
chiefs_runpassonly = chiefs[chiefs['play_type'].isin(['run','pass'])]

In [10]:
nfl_runpassonly = nfl[nfl['play_type'].isin(['run','pass'])]

In [11]:
pd.crosstab(chiefs_runpassonly.down,chiefs_runpassonly.play_type,margins = True, normalize='index')

play_type      pass       run
down                         
1.0        0.552239  0.447761
2.0        0.644860  0.355140
3.0        0.760000  0.240000
4.0        0.400000  0.600000
All        0.617347  0.382653

In [12]:
pd.crosstab(nfl_runpassonly.down,nfl_runpassonly.play_type,margins = True, normalize='index')

play_type      pass       run
down                         
1.0        0.506966  0.493034
2.0        0.592352  0.407648
3.0        0.785636  0.214364
4.0        0.639098  0.360902
All        0.593548  0.406452

> **Finding: Chiefs are more likely to call Pass Plays for all downs (61.7% vs 59.4%). Let's now look at play data based on down and distance**

In [13]:
chiefs_temp = pd.crosstab(chiefs_runpassonly.ydstogo_label,[chiefs_runpassonly.play_type,chiefs_runpassonly.down],normalize='index')

In [14]:
chiefs_temp2 = pd.crosstab(chiefs_runpassonly.ydstogo_label,chiefs_runpassonly.down,normalize='index')

In [15]:
chiefs_crosstab_play=chiefs_temp/chiefs_temp2

In [16]:
nfl_temp = pd.crosstab(nfl_runpassonly.ydstogo_label,[nfl_runpassonly.play_type,nfl_runpassonly.down],normalize='index')

In [17]:
nfl_temp2 = pd.crosstab(nfl_runpassonly.ydstogo_label,nfl_runpassonly.down,normalize='index')

In [18]:
nfl_crosstab_play=nfl_temp/nfl_temp2

In [19]:
chiefs_crosstab_play 

play_type          pass                                     run            \
down                1.0       2.0       3.0       4.0       1.0       2.0   
ydstogo_label                                                               
<2             0.300000  0.416667  0.513514  0.272727  0.700000  0.583333   
2-5            0.428571  0.615385  0.842105  0.500000  0.571429  0.384615   
6-9            0.583333  0.686047  0.734694  1.000000  0.416667  0.313953   
10             0.555012  0.670455  0.954545       NaN  0.444988  0.329545   
11-14          0.500000  0.735294  0.809524       NaN  0.500000  0.264706   
15+            0.909091  0.750000  1.000000       NaN  0.090909  0.250000   

play_type                          
down                3.0       4.0  
ydstogo_label                      
<2             0.486486  0.727273  
2-5            0.157895  0.500000  
6-9            0.265306  0.000000  
10             0.045455       NaN  
11-14          0.190476       NaN  
15+            0.000000       NaN

In [20]:
nfl_crosstab_play

play_type          pass                                     run            \
down                1.0       2.0       3.0       4.0       1.0       2.0   
ydstogo_label                                                               
<2             0.304933  0.347505  0.432304  0.406475  0.695067  0.652495   
2-5            0.452229  0.488372  0.854447  0.857143  0.547771  0.511628   
6-9            0.397015  0.640568  0.889070  0.893939  0.602985  0.359432   
10             0.506940  0.625427  0.889085  0.884615  0.493060  0.374573   
11-14          0.620968  0.753029  0.898305  1.000000  0.379032  0.246971   
15+            0.716814  0.722513  0.804825  1.000000  0.283186  0.277487   

play_type                          
down                3.0       4.0  
ydstogo_label                      
<2             0.567696  0.593525  
2-5            0.145553  0.142857  
6-9            0.110930  0.106061  
10             0.110915  0.115385  
11-14          0.101695  0.000000  
15+            0.195175  0.000000

In [22]:
chiefs_crosstab_play-nfl_crosstab_play

play_type          pass                                     run            \
down                1.0       2.0       3.0       4.0       1.0       2.0   
ydstogo_label                                                               
<2            -0.004933  0.069162  0.081209 -0.133748  0.004933 -0.069162   
2-5           -0.023658  0.127013 -0.012342 -0.357143  0.023658 -0.127013   
6-9            0.186318  0.045478 -0.154376  0.106061 -0.186318 -0.045478   
10             0.048072  0.045028  0.065461       NaN -0.048072 -0.045028   
11-14         -0.120968 -0.017735 -0.088781       NaN  0.120968  0.017735   
15+            0.192277  0.027487  0.195175       NaN -0.192277 -0.027487   

play_type                          
down                3.0       4.0  
ydstogo_label                      
<2            -0.081209  0.133748  
2-5            0.012342  0.357143  
6-9            0.154376 -0.106061  
10            -0.065461       NaN  
11-14          0.088781       NaN  
15+           -0.195175       NaN

> **Summary of play pass tendancy on more common down and distance scenarios:**

| Down & Distance | Chiefs Pass % | NFL Pass % |
| --------------- | ----------- | ----------- |
| 1st & 10   | 55.5%    | 50.6% |
| 2nd & 10	 | 67.0% | 62.5% | 
| 2nd & 5    | 61.5% | 48.8% | 
| 2nd & 1	 | 41.7% | 34.7% | 
| 3rd & 10	 | 95.5% | 88.9% | 
| 3rd & 5    | 84.2% | 85.4% | 
| 3rd & 1	 | 51.3% | 43.2% | 

The chiefs call more pass plays than the average NFL team for all down/distance scenarios exccept 3rd & 5

2nd down & 5 yards is a scenario the Chiefs like to call pass plays and they do so at a much higher rate than the average NFL team (Chiefs 61.5% vs NFL team of 48.8%) 

Run plays are more likely in short yardage especially on 3rd down scenarios. However, even in 3rd and short the Chiefs are more likely to call pass plays (51%). 



## Step 4: Which NFL teams are the most pass-heavy

In [29]:
pd.crosstab(nfl_runpassonly.posteam,nfl_runpassonly.play_type,margins = True, normalize='index').sort_values(by='run')

play_type      pass       run
posteam                      
PIT        0.682646  0.317354
GB         0.677483  0.322517
ATL        0.658052  0.341948
MIN        0.650602  0.349398
NYG        0.643655  0.356345
TB         0.636711  0.363289
IND        0.630975  0.369025
CIN        0.620022  0.379978
PHI        0.619279  0.380721
KC         0.617108  0.382892
DEN        0.616152  0.383848
OAK        0.614754  0.385246
ARI        0.611857  0.388143
DET        0.609082  0.390918
CLE        0.605523  0.394477
CAR        0.594567  0.405433
All        0.594171  0.405829
JAX        0.588766  0.411234
LAC        0.588047  0.411953
SF         0.584089  0.415911
MIA        0.582851  0.417149
NYJ        0.581761  0.418239
DAL        0.577535  0.422465
LA         0.576369  0.423631
WAS        0.574268  0.425732
NE         0.563810  0.436190
HOU        0.549268  0.450732
CHI        0.547643  0.452357
NO         0.544995  0.455005
BUF        0.540080  0.459920
BAL        0.526032  0.473968
TEN        0.520474  0.479526
SEA        0.474112  0.525888

> **The Chiefs are the 10th most pass heavy team in the NFL**